src/ImageSharp.Web.Providers.AWS/Providers/AWSS3StorageImageProvider.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.Runtime;
+using Amazon.S3;
+using Amazon.S3.Model;
+using Amazon.Util;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using SixLabors.ImageSharp.Web.Resolvers;
+
+namespace SixLabors.ImageSharp.Web.Providers
+{
+    /// <summary>
+    /// Returns images stored in AWS S3.
+    /// </summary>
+    public class AWSS3StorageImageProvider : IImageProvider
+    {
+        /// <summary>
+        /// Character array to remove from paths.
+        /// </summary>
+        private static readonly char[] SlashChars = { '\\', '/' };
+
+        private readonly IAmazonS3 amazonS3Client;
+        private readonly ILogger<AWSS3StorageImageProvider> logger;
+        private readonly AWSS3StorageImageProviderOptions storageOptions;
+        private Func<HttpContext, bool> match;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AWSS3StorageImageProvider"/> class.
+        /// </summary>
+        /// <param name="amazonS3Client">Amazon S3 client</param>
+        /// <param name="logger">Microsoft.Extensions.Logging ILogger</param>
+        /// <param name="storageOptions">The S3 storage options</param>
+        public AWSS3StorageImageProvider(IAmazonS3 amazonS3Client, ILogger<AWSS3StorageImageProvider> logger, IOptions<AWSS3StorageImageProviderOptions> storageOptions)
+        {
+            Guard.NotNull(storageOptions, nameof(storageOptions));
+
+            this.amazonS3Client = amazonS3Client;
+            this.logger = logger;
+            this.storageOptions = storageOptions.Value;
+        }
+
+        /// <inheritdoc />
+        public Func<HttpContext, bool> Match
+        {
+            get => this.match ?? this.IsMatch;
+            set => this.match = value;
+        }
+
+        /// <inheritdoc />
+        public bool IsValidRequest(HttpContext context)
+        {
+            var displayUrl = context.Request.Path;
+            return Path.GetExtension(displayUrl).EndsWith(".jpg") || Path.GetExtension(displayUrl).EndsWith(".png");
+        }
+
+        /// <inheritdoc />
+        public async Task<IImageResolver> GetAsync(HttpContext context)
+        {
+            PathString displayUrl = context.Request.Path;
+            this.logger.LogDebug("Getting image for {ImageUri}", displayUrl);
+            string imageId = Path.GetFileNameWithoutExtension(displayUrl);
+            this.logger.LogDebug("Image id is {ImageId}", imageId);
+
+            string imagePath = $"{imageId}.png";
+
+            bool imageExists = await this.KeyExists(this.storageOptions.BucketName, imagePath);
+
+            if (!imageExists)
+            {
+                this.logger.LogDebug("No image found for {ImageId}", imageId);
```

Comment:
Don't need these logger statements here. We log in the middleware.

Reply:
Too easy will clean these up.